### ISSUE TITLE
🦠 Model Request: interpretable prediction of drug-induced liver injury through permutation feature importance and attention mechanism

### ISSUE THREAD

**Issue by leilayesufu** [ersilia-os/ersilia#947]
### Model Name

interpretable prediction of drug-induced liver injury through permutation feature importance and attention mechanism

### Model Description

This study focused on interpreting the model prediction to analyze which features could potentially cause DILI and indicated that the model not only captured the previously known properties to be related to DILI but also proposed a new DILI potential substructural of physicochemical properties.

### Slug

InterDILI

### Tag

Toxicity, Human, Cardiotoxicity

### Publication

https://jcheminf.biomedcentral.com/articles/10.1186/s13321-023-00796-8

### Source Code

https://github.com/bmil-jnu/InterDILI/tree/main

### License

None

**Comment by leilayesufu** [comment]
@GemmaTuron @DhanshreeA 
This is the second Dili model @miquelduranfrigola shared in the [#literature](https://ersilia-workspace.slack.com/archives/C010067BP2Q) channel. I have read the publication and gone through the source code. 
The dataset is given and the code to change from smiles to the two features used. 
 https://github.com/bmil-jnu/InterDILI/tree/main/Molecular%20Descriptors

Four machine learning models were constructed and optimized:

Random Forest (RF)
Light Gradient Boosting Machine (LGBM)
Logistic Regression (LR)
Neural Network with Attention (NN with attention)

**Comment by miquelduranfrigola** [comment]
OK thanks @leilayesufu this is useful.

Two comments:
- We do not need to have the "interpretability" capabilities necessarily. I think it is fine if we just predict DILI outcome.
- Have you been able to run their source code? Do they provide pre-trained parameters?

Thanks!

PS: @GemmaTuron feel free to approve this model request.

**Comment by leilayesufu** [comment]
Thank you! @miquelduranfrigola 
They do not provide the pre trained checkpoints, just the model datasets so we’d have to retrain the model with the dataset

**Comment by miquelduranfrigola** [comment]
OK perfect, thanks @leilayesufu

**Comment by leilayesufu** [comment]
@GemmaTuron @miquelduranfrigola @DhanshreeA 
Good morning. In this model after reading the publication and going through the source code. 
The authors provided the training dataset. https://github.com/bmil-jnu/InterDILI/blob/main/Data%20Collection/Total_dataset.csv
```
Number of Negative (0) outcomes: 898
Number of Positive (1) outcomes: 952
Total number of samples: 1850
```
and the code to convert it to the two features they used. https://github.com/bmil-jnu/InterDILI/blob/main/Molecular%20Descriptors/Smiles%20to%20Features.py

They trained the features with Random forest classifier, LightGBM model, Logistic regression and  MLPClassifier 

The RF performed the best according to the authors with 
```
**Hold-out Validation:**
Accuracy: Ranged from 0.81 to 0.90.
Sensitivity: Ranged from 0.77 to 0.96.
Specificity: Ranged from 0.62 to 0.90.
Precision: Ranged from 0.74 to 0.85.
F1 Score: Ranged from 0.77 to 0.89.
AUROC (Area Under the Receiver Operating Characteristic Curve): Ranged from 0.88 to 0.97.
AUPRC (Area Under the Precision-Recall Curve): Ranged from 0.81 to 0.95.
**Stratified k-fold Cross Validation:**
Accuracy: Ranged from 0.72 to 0.78.
Sensitivity: Ranged from 0.71 to 0.80.
Specificity: Ranged from 0.72 to 0.76.
Precision: Ranged from 0.73 to 0.78.
AUROC: Ranged from 0.76 to 0.87.
AUPRC: Ranged from 0.75 to 0.87.
```

I trained the model and saved the five splits and i got RF to be the best performing with auroc of 0.85 and 
```
Average Classification Report Across All Folds:

              precision    recall  f1-score   support

           0       0.74      0.76      0.75       871
           1       0.78      0.76      0.77       979

    accuracy                           0.76      1850
   macro avg       0.76      0.76      0.76      1850
weighted avg       0.76      0.76      0.76      1850
```
I have created a pull request. 
https://github.com/ersilia-os/eos21q7/pull/1

**Comment by miquelduranfrigola** [comment]
This looks very good @leilayesufu ! Thanks so much for a very detailed update.

**Comment by DhanshreeA** [comment]
@leilayesufu Fantastic work Leila, thanks for the updates!

@miquelduranfrigola it might be fun to see if we can reproduce the author's results with the DILIRank dataset using the model that Leila has trained. What do you think?

In any case, we should run this model against the DILI test set from TDC and see what we get?

**Comment by leilayesufu** [comment]
@GemmaTuron This model is ready for review. https://github.com/ersilia-os/eos21q7

**Comment by GemmaTuron** [comment]
Hi @leilayesufu 

Thanks for the work, the notebooks are much easier to read and follow, and there is less redundant code!
I have a few questions before we can consider the model incorporated:
1. The authors state that they use both Descriptors and Physicochemical properties as training data (_using substructure and physicochemical descriptors as inputs and the DILI label as the output_). I see you only took the descriptors, so the train_test validation should be repeated using both (collating them in a single input file)
2. Once you do that, use the exact code they provide in the smiles to features file on the main.py to get the features, make sure to collate them in the same order as in the training
3. You are saving a model that corresponds to one  of the splits, please, always train a final model with ALL the data and save that one
4. Do not save the train test splits with the descriptors, simply collate them when necessary. See how I have done it for the other DILI model

Let's make sure we get this right before moving onto comparing the models!

**Comment by GemmaTuron** [comment]
I suggest, instead of train:test splits, we follow exactly what the authors have done:
they take 4 datasets as train, and one as test - let's see if we can validate the results using a RF - we can do that only for this model, since the authors already report is the best one

**Comment by GemmaTuron** [comment]
Hi @leilayesufu 

The dataset has a column that indicates the origin of the molecule - we can use that as separator
For the descriptors, as they are numerical descriptors I understand they are putting them together in a single array or dataframe

**Comment by leilayesufu** [comment]
@GemmaTuron The authors mentioned that in the dataset DILIrank was used for testing while the others was used as training, so i used this division in the train test split and i have divided it into train and test, i used the smiles to features to convert it to both descriptors and physicochemical descriptors and i trained based on that and i have updated the main.py file to use the final model with the random forest classifier model , i have created a pull request
<img width="940" alt="datase" src="https://github.com/ersilia-os/ersilia/assets/78659709/91ddb189-f13b-4aab-8109-8df4782ce5ee">

**Comment by GemmaTuron** [comment]
Hi @leilayesufu 

Before merging the pull request can you summarize how this model performs compared to the previous test?

**Comment by leilayesufu** [comment]
When using just the descriptors with the train test splits and random forest classifier I got an auroc of 0.85
While when using the molecules the author used for train and test with both descriptors and pysiochemical descriptors I get an auroc of 0.97

**Comment by GemmaTuron** [comment]
great, this sounds mode like what the authors were reporting: receiver operating characteristic (AUROC) of 0.88–0.97 
I've merged the PR

**Comment by GemmaTuron** [comment]
Hi @leilayesufu

I have made modifications to your code. Most important - did you carefully read the publication before going into model training? the authors clearly state that they scale the physicochemical descriptors - please read the methods section of the paper.
I don't see this done anywhere in your code, and it is crucial.
I have refactored the Data_processing notebook, reducing duplicated code - no need to run it again
I have started preparing the model_training notebook, starting by creating functions to generate the molecule descriptors - you'll see I've made three functions, one for ecfp, one for qed and one for scaling, which I've left for you to complete. Then create a fourth function that uses the other three to generate the data in the format we want.
Once this is done, you can retrain the model using the new code. No need to save the descriptor files, simply call the functions to calculate them and pass them directly to the model.
Also, please use updated version of python (py3.11) - make sure to upgrade package versions as well to the newest, since we are retraining the full model - and pin them in the Dockerfile

Hope this is clear but let me know otherwise.

**Comment by GemmaTuron** [comment]
In the supplementary figures of the paper, the authors show the distribution of the physchem descriptors before and after the scaling so I suggest we do the same check to compare - supplementary information can be found at the bottom of the publication link in a downloadable .docx file

**Comment by leilayesufu** [comment]
@GemmaTuron Good afternoon
I have updated the dockerfile to use python  (py3.11). I have also updated the model_training notebook scaling the qed. In the model training i also Plotted the histograms of descriptors before scaling and after scaling.
I have also updated the main.py and created a PR

**Comment by GemmaTuron** [comment]
Hi @leilayesufu 

Can you explain how have you scaled the features, taking into account what the authors did? also the plots are good but please make them separte (with and without scaling) because I cannot see very well the scaled values.
Once this is clear, to make the file more readable, create a script with the functions (maybe with a Class?) and then import it into main.py, instead of pasting all the functions there

**Comment by leilayesufu** [comment]
In the preprocessing step, the authors scaled the physicochemical descriptors to zero mean and unit variance using the StandardScaler from scikit-learn. This ensures that each feature has a mean of 0 and a standard deviation of 1 after scaling.
In the provided code, after calculating the QED descriptors, the StandardScaler is applied to scale these descriptors. Here's how i did it.
```
# Function to scale QED descriptors
def scale_qed(df):
    scaler = StandardScaler()
    scaled_array = scaler.fit_transform(df)
    scaled_df = pd.DataFrame(scaled_array, columns=df.columns)
    return scaled_df
```

 StandardScaler is instantiated, and then the fit_transform method is called on the descriptors (qed_df). This method first fits the scaler to the data (calculating the mean and standard deviation), and then transforms the data using the calculated mean and standard deviation.

After scaling, the descriptors have zero mean and unit variance.

I have seperated the plots. so it has plots before scaling and after scaling.

**Comment by leilayesufu** [comment]
I also updated the main.py so it imports a class with all the functions. the code works well locally when testing with bash. but i am getting this error with ersilia.

```
StandardModelExampleError
Detailed error:
Standard model run from CSV was not possible for model eos21q7
Output file /home/runner/eos/dest/eos21q7/example_standard_output.csv was not created successfully
Hints:
If you fetch this model from Docker Hub, or you are running it through URL, this is the first time run is executed in your local computer. Reach out to Ersilia to get specific help.
[]
🚨🚨🚨 Something went wrong with Ersilia 🚨🚨🚨
Error message:
Ersilia exception class:
StandardModelExampleError
Detailed error:
Standard model run from CSV was not possible for model eos21q7
Output file /home/runner/eos/dest/eos21q7/example_standard_output.csv was not created successfully
Hints:
If you fetch this model from Docker Hub, or you are running it through URL, this is the first time run is executed in your local computer. Reach out to Ersilia to get specific help.
If this error message is not helpful, open an issue at:
 - https://github.com/ersilia-os/ersilia
Or feel free to reach out to us at:
 - hello[at]ersilia.io
If you haven't, try to run your command in verbose mode (-v in the CLI)
 - You will find the console log file in: /home/runner/eos/current.log
Error: Process completed with exit code 1.
```

**Comment by GemmaTuron** [comment]
@leilayesufu 
Let's fix the standard scaler with the parameters set to false so we get a mean of 0 and variance of 1: https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html

**Comment by leilayesufu** [comment]
@GemmaTuron I have been doing some reading and i read that By default, with_mean=True and with_std=True, so the StandardScaler will center the data around the mean (making it 0) and scale it by the standard deviation, resulting in a variance of 1 for the scaled data.

https://scikit-learn.org/stable/auto_examples/preprocessing/plot_all_scaling.html#standardscaler

> with_mean=False: If you set this parameter to False, it means that the scaler will not subtract the mean from the data during scaling. Instead, it will use a mean of zero for scaling. So, if with_mean=False, u would be zero.
> 
> with_std=False: If you set this parameter to False, it means that the scaler will not divide the data by the standard deviation during scaling. Instead, it will use a standard deviation of one for scaling. So, if with_std=False, s would be one.
> 
> Therefore, if you initialize the StandardScaler with both with_mean=False and with_std=False, it will effectively perform no scaling, leaving the data unchanged, as it doesn't subtract the mean nor divide by the standard deviation during scaling

The standard score of a sample x is calculated as:

z = (x - u) / s

where u is the mean of the training samples or zero if with_mean=False, and s is the standard deviation of the training samples or one if with_std=False.
By default, StandardScaler centers the data (i.e., sets the mean to 0) and scales it (i.e., sets the standard deviation to 1) unless you explicitly set with_mean=False or with_std=False. Therefore, we don't need to explicitly specify these parameters if you want mean 0 and variance 1. So i think the default scales it to a mean of 0 and variance of 1.

https://stackoverflow.com/questions/40758562/can-anyone-explain-me-standardscaler

If you set with_mean=False and/or with_std=False when instantiating the StandardScaler object, it changes the behavior of the scaler as follows:

with_mean=False: This means that the scaler will not center the data by subtracting the mean. Instead, it assumes that the data is already centered around 0. In this case, the resulting scaled data will not have a mean of 0.
with_std=False: This means that the scaler will not scale the data to have unit variance. Instead, it assumes that the data already has unit variance. In this case, the resulting scaled data will not have a variance of 1.

<img width="728" alt="scaler" src="https://github.com/ersilia-os/ersilia/assets/78659709/516d10b9-6602-4cfc-8a78-ad053fb2ffcc">

**Comment by leilayesufu** [comment]
https://www.digitalocean.com/community/tutorials/standardscaler-function-in-python

**Comment by GemmaTuron** [comment]
Good reading @leilayesufu, maybe on a quick look I was confused, sorry!
In any case, let's be sure as this is easy to check: on the scaled data, check that the mean is actually 0 in each column and the variance is 1, you can use the pandsas functions mean and std for that.
After this, make sure to save the right fitted scaler and only apply it to transform (not fit) predicted data.
Thanks!

**Comment by leilayesufu** [comment]
@GemmaTuron 
Good morning. i  have saved the fitted scaler in the checkpoints folder.  We now have two checkpoints the model pickle file and the scaler pickle file, i have also updated the code so that it transform with the saved scaler. 
Also, i have checked for the mean and the variance  of the processed scaled qed descriptors with the following code

```
## Check that the mean  is actually 0 in each column and the variance is 1 [for the qed descriptors]
# Filter out the specified columns
selected_columns = ['MW', 'ALOGP', 'HBA', 'HBD', 'PSA', 'ROTB', 'AROM', 'ALERTS']
selected_descriptors = processed_train_data[selected_columns]

# Calculate the mean and variance for each column
mean_values_selected = selected_descriptors.mean()
variance_values_selected = selected_descriptors.var()

print("Mean of selected descriptors:")
print(mean_values_selected)

print("\nVariance of selected descriptors:")
print(variance_values_selected)
```

I got
```
Mean of selected descriptors:
MW       -1.728072e-16
ALOGP    -1.143577e-16
HBA      -1.016513e-16
HBD       0.000000e+00
PSA       7.623849e-18
ROTB      2.541283e-17
AROM     -4.066053e-17
ALERTS   -1.016513e-16
dtype: float64

Variance of selected descriptors:
MW        1.000716
ALOGP     1.000716
HBA       1.000716
HBD       1.000716
PSA       1.000716
ROTB      1.000716
AROM      1.000716
ALERTS    1.000716
dtype: float64
```

The means are centered around zero and their variances close to one. which indicates that the selected descriptors have been successfully standardized. I have  created a PR

**Comment by GemmaTuron** [comment]
Great @leilayesufu good work!

I've merged the PR

**Comment by GemmaTuron** [comment]
This model is completed!
